pandas/core/groupby/groupby.py

```diff
@@ -689,7 +689,22 @@ def _set_result_index_ordered(self, result):
             result.set_axis(index, axis=self.axis, inplace=True)
             result = result.sort_index(axis=self.axis)
 
-        result.set_axis(self.obj._get_axis(self.axis), axis=self.axis, inplace=True)
+        return self._restore_grouper_index(result)
+
+    def _restore_grouper_index(self, result: FrameOrSeries) -> FrameOrSeries:
+        # GH 35612
+        # result.index is a standard index => may need to restore original index
+
+        rows_dropped: bool = len(result) < len(self._selected_obj)
+        # TODO: address case of axis==1, dropped columns
+
+        if self.dropna and not self.axis and rows_dropped:
+            # use result.index to select from index of original object
+            original_index = self._selected_obj.index[result.index]
```

Comment:
how is this an *indexer* into the original index?

Reply:
So AFAIK after factorize `result` has a simple index, meaning if the row indices originally were `['a', 'b', 'c']` and, say, `'b'` was dropped in factorization, `result.index` at the top of this method will be `[0, 2]`. We need to restore the original index to the transformed groupby result ergo this slice op.

On the current master in `SeriesGroupBy._transform_general` the index is reset at the end:
https://github.com/pandas-dev/pandas/blob/e975f3def1ff430d5801fbe241c52d7206c79956/pandas/core/groupby/generic.py#L560
but this part fails when rows get dropped b/c `result.index` and `self._selected_obj` no longer have matching shape